### PR TITLE
TXM In-memory: add priority queue: STEP 1

### PR DESCRIPTION
**Pull request by poopoothegorilla**
This PR is the start of a chain of PRs to make the review process easier. The PR adds a priority queue that will be used by the in-memory TXM.

NOTES:
- utilizes `container/heap` pkg https://pkg.go.dev/container/heap

### CODE REVIEW

File: common/txmgr/priority_queue.go
Comment (jmank88) [inline common/txmgr/priority_queue.go:52]: Would it make sense to capture these low level ops as a separate clean generic heap, without the tx queue specifics? Then this can just have a field of type `Heap[*txmgrtypes.Tx[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, SEQ, FEE]` without mixing in explicit casts.

File: common/txmgr/priority_queue.go
Comment (patrick-dowell) [inline common/txmgr/priority_queue.go:21]: I'm wondering a little why we need a TxPriorityQueue instead of just accessing the PriorityQueue directly. I see that the mutex is in the TxPriorityQueue, though we could move that logic to the PriorityQueue to ensure that anyone who uses it would have a thread-safe PriorityQueue. And there are two higher level functions (RemoveTxById) and PruneByTxId that wouldn't entirely make sense to be built into the PriorityQueue. But the FindById function is in the PriorityQueue interface, and that's the main thing this functions rely on, so in theory, you could just put all this in PriorityQueue and get rid of TxPriorityQueue right? Unless I'm missing something.

**Review by patrick-dowell — COMMENTED** [review]
Overall this looks reasonable, but I notice a lot of redundant code, mainly I think because TxPriorityQueue is thinly wrapping PriorityQueue. I wonder if it'd be a little cleaner to just collapse these two files together and save some code lines / complexity.

File: common/txmgr/priority_queue.go
Comment (amit-momin) [inline common/txmgr/priority_queue.go:85]: Just a small optimization. Couldn't we use read lock here?
```suggestion
	pq.RLock()
	defer pq.RUnlock()
```

File: common/txmgr/types/priority_queue.go
Comment (prashantkumar1982) [inline common/txmgr/types/priority_queue.go:10]: This type seems internal to the TXM right?
I mean not exposed to anywhere else outside the TXM.
Could you not export it (rename to priorityQueue) and move it out of types package?
The types package I think is for types we want to export to other packages.

File: common/txmgr/priority_queue.go
Comment (prashantkumar1982) [inline common/txmgr/priority_queue.go:19]: You don't need to make the change, but just wondering out aloud.
The only pattern we have need for, is add items to end of queue, and remove items from the beginning.
And at times, search an item in this list, and remove it.

Could we not be better served instead by a doubly liked list data structure?
In Linked List, insertions and delete from head would be O(1), and deletions by id would be O(n).

Just a thought for future. Not to do this right away

File: common/txmgr/priority_queue.go
Comment (prashantkumar1982) [inline common/txmgr/priority_queue.go:12]: UnExport this too.
Likely this is internal to TXM.

**Review by prashantkumar1982 — COMMENTED** [review]
I think this is good to go.
Please add unit-tests and mark this PR as Ready for Review.
And lets merge it as soon as you get 2 approvals.

File: common/txmgr/priority_queue.go
Comment (prashantkumar1982) [inline common/txmgr/priority_queue.go:29]: The maxUnstarted name doesn't make sense inside this file.
Rename it to capacity or size.

File: common/internal/queues/tx_priority_queue_test.go
Comment (prashantkumar1982) [inline common/internal/queues/tx_priority_queue_test.go:1]: nice job on being thorough!

**Review by prashantkumar1982 — DISMISSED** [review]
This looks pretty good.
Please send it out to #bix-code-review separately too, for a final reminder to review. And then plz merge this by end of day tomorrow with whatever approvals you get.

File: common/internal/queues/tx_priority_queue.go
Comment (prashantkumar1982) [inline common/internal/queues/tx_priority_queue.go:29]: Do we want to panic or return null if capacity == 0?

**Review by amit-momin — DISMISSED** [review]

**Review by amit-momin — APPROVED** [review]

File: common/internal/queues/priority_heap.go
Comment (patrick-dowell) [inline common/internal/queues/priority_heap.go:135]: Given that go is garbage collected, wouldn't this get cleaned up after line 136? I would imagine the older slice gets dealloc and the newer slice doesn't have a pointer to old[n-1] anymore. I imagine there's something going on here that I don't quite understand.

File: common/internal/queues/priority_heap.go
Comment (patrick-dowell) [inline common/internal/queues/priority_heap.go:134]: Isn't this going to panic when n = 0?

File: common/internal/queues/tx_priority_queue_test.go
Comment (patrick-dowell) [inline common/internal/queues/tx_priority_queue_test.go:18]: It'd be good to have more corner case testing for the priority queue. Also are there no tests for the priority heap?

File: common/internal/queues/priority_heap.go
Comment (patrick-dowell) [inline common/internal/queues/priority_heap.go:104]: I think we need to check to make sure i & j are valid indices.

File: common/internal/queues/priority_heap.go
Comment (patrick-dowell) [inline common/internal/queues/priority_heap.go:110]: same as above

File: common/internal/queues/priority_heap.go
Comment (patrick-dowell) [inline common/internal/queues/priority_heap.go:126]: shouldn't this be len() - 1?

File: common/internal/queues/tx_priority_queue.go
Comment (patrick-dowell) [inline common/internal/queues/tx_priority_queue.go:37]: nit: I think it'd make the code slightly more readable to give the internal heap a better name, like `p_heap` or something. Because it reflects internal state and is defined near the top of the file whereas `pq` is defined on the line above and is easy to check and see what it is.

File: common/internal/queues/tx_priority_queue.go
Comment (patrick-dowell) [inline common/internal/queues/tx_priority_queue.go:56]: You know, I went searching through the code and I can't find where Remove is defined on the heap. I'm curious why the index is off by 1 - that seems unusual.

**Review by patrick-dowell — COMMENTED** [review]
Left a bunch of comments.

File: common/internal/queues/priority_heap.go
Comment (dhaidashenko) [inline common/internal/queues/priority_heap.go:121]: This increases complexity to add N elements into queue to N^2 instead of NlogN.
Push should simply append new element to the slice.
Cap restrictions should be performed on the upper layer. AddTx will look like
```
if pq.ph.Len() == maxSize {
_ = heap.Pop(pq.ph)
}

heap.Push(pq.ph, tx)
```

File: common/internal/queues/tx_priority_queue.go
Comment (dhaidashenko) [reply common/internal/queues/tx_priority_queue.go:56]: That's because of `pq.idToIndex[tx.(*txmgrtypes.Tx[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, SEQ, FEE]).ID] = len(pq.txs)` instead of `pq.idToIndex[tx.(*txmgrtypes.Tx[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, SEQ, FEE]).ID] = len(pq.txs)-1`
+1 on fixing this

File: common/internal/queues/priority_heap.go
Comment (dhaidashenko) [inline common/internal/queues/priority_heap.go:50]: setting to nil is sufficient, no need to iterate through all of the elements to set them nil

File: common/internal/queues/priority_heap.go
Comment (dhaidashenko) [inline common/internal/queues/priority_heap.go:57]: Nit: return (0, false) instead of (-1) to make it consistent with go libs

**Review by dhaidashenko — CHANGES_REQUESTED** [review]

**Review by dhaidashenko — COMMENTED** [review]
Changing status to "Comment" instead of "Requested changes" to not block merge if it needs to go through. Requested changes are important, but manly affect readability and performance of internal component

File: common/internal/queues/priority_heap.go
Comment (dhaidashenko) [inline common/internal/queues/priority_heap.go:67]: nit: cap and len are safe on nil slices

File: common/internal/queues/tx_priority_queue.go
Comment (dhaidashenko) [inline common/internal/queues/tx_priority_queue.go:71]: nit: `_` looks like python's notation. Small letter should be sufficient to signal that method is private

**Review by dhaidashenko — DISMISSED** [review]
LGTM

File: .changeset/seven-rice-end.md
Comment (prashantkumar1982) [inline .changeset/seven-rice-end.md:1]: Just curious, why this file name?

**Review by prashantkumar1982 — COMMENTED** [review]
LGTM, can approve once you are done with all comments!

File: common/internal/queues/priority_heap.go
Comment (patrickhuie19) [inline common/internal/queues/priority_heap.go:30]: /nit `if capacity >= 0` ?

File: common/internal/queues/priority_heap.go
Comment (patrickhuie19) [inline common/internal/queues/priority_heap.go:10]: Re: concurrency safe, what if `Swap` and `Pop` are called concurrently?

File: common/internal/queues/priority_heap.go
Comment (patrickhuie19) [reply common/internal/queues/priority_heap.go:30]: Does it matter that the consumer can specify a negative capacity?

File: common/internal/queues/priority_heap.go
Comment (patrickhuie19) [reply common/internal/queues/priority_heap.go:10]: I don't understand. To me it looks like you've taken something that is not concurrent safe `priorityHeap` and made it concurrent safe in `TxPriorityQueue` by adding a locking mechanism (the RWMutex).

File: common/internal/queues/priority_heap.go
Comment (patrickhuie19) [inline common/internal/queues/priority_heap.go:56]: /nit is this inconsistent? else where in `priorityHeap` the pattern is to have the caller responsible for providing appropriate inputs based on the state of the queue. Here you sanity check for the caller that the length of the queue is not 0.

File: common/internal/queues/tx_priority_queue.go
Comment (patrickhuie19) [inline common/internal/queues/tx_priority_queue.go:43]: Restating my understanding:

`priorityHeap` is meant to store tx ordered by creation time, with the oldest tx at index 0, and the newest tx at index length - 1. 

The current implementation of Pop on `priorityHeap`:

```
func (pq *priorityHeap[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, R, SEQ, FEE]) Pop() any {
	old := pq.txs
	n := len(old)
	tx := old[n-1]
	old[n-1] = nil // avoid memory leak
	pq.txs = old[0 : n-1]
	delete(pq.idToIndex, tx.ID)
	return tx
}
```

Here, IFIUC, we remove the tx at index length - 1 from both the array and the index map, which is the tx with the most recent creation time. 

Then, in `AddTx`, we have 

```
	if pq.ph.Len() == pq.ph.Cap() {
		heap.Pop(pq.ph)
        }
```	

So, when the queue is full, doesn't the current code remove the newest transaction?

**Review by dhaidashenko — DISMISSED** [review]

**Pull request closed without merging**
